<a href="https://colab.research.google.com/github/tt35/AI_Tranformers_vs_Fastai/blob/main/Final_Project_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Transformers vs. Fastai in image classification task - Part3**
### Which one is better and more suitable for learning?

In this notebook, I would like to compare the accuracy and easiness of the Transformers model with the Fast.ai model.

## Fast.ai model

The code below makes the Fast.ai image classfier by using the same images used for the Transformers model. Moreover, the model performance was evaluated with the same images as well.

In [ ]:
from fastbook import *

In [ ]:
path = Path('/home/tt35/Desktop/pics_fastai_train/')

In [ ]:
def is_CF(x):
    return x[0] == "C"

image_files = get_image_files(path).sorted()

set_seed(1000)
dls = ImageDataLoaders.from_name_func(
    path, image_files,
    valid_pct=0.2, seed=42,
    label_func=is_CF,
    item_tfms=Resize(224),
    bs=2)

In [ ]:
learn = cnn_learner(
    dls=dls,
    arch=resnet18,
    metrics=accuracy,
)
learn.fine_tune(epochs=3)

In [ ]:
test_path = Path('/home/tt35/Desktop/pics_fastai_test/')
test_image_files = get_image_files(test_path)

In [ ]:
test_dl = dls.test_dl(test_image_files, with_labels=True)

In [ ]:
interp = ClassificationInterpretation.from_learner(learn, dl=test_dl)

In [ ]:
accuracy(interp.preds, interp.targs)

TensorBase(0.9412)

## Results (accuracy on the test data, epoch = 3)

**Fast.ai Model**
```
seed = 1: 
  accuracy = 0.8235
  
seed = 10:
  accuracy = 0.7353
  
seed = 100:
  accuracy = 0.7647
  
seed = 1000:
  accuracy = 0.9412
  ```

**Transformers Model**
```
seed = 1: 
  eval_accuracy = 1.0
 
seed = 10:
  eval_accuracy = 1.0
  
seed = 100:
  eval_accuracy = 1.0
  
seed = 1000:
  eval_accuracy = 0.8824
  ```

## Reflection

The range of accuracy for Transformers model is 0.88 to 1.0, and it is 0.74 to 0.94. As the results show, the accuracy of Transformers model is much better than the one of Fast.ai model. In terms of usability, I would say Fast.ai model was a lot easier as it has some very useful API such as ImageDataLoaders.from_path_func. Unlike Fast.ai, Transformers do not have straightforward APIs, which requires the users to write a lot of code on their own.

## Limitations and Future Directions	

As I said above, the limitations come from the variety of pictures. For example, some images include cars and gloomy sky. Since I am not exactly sure how much these factors influence the model performance, the future direction is to create attention heat map, so that we can see exactly what parts the model pays attention. 